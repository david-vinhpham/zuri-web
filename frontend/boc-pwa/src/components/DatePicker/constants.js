import moment from 'moment/moment';

export const THIS_YEAR = moment().year();
export const THIS_MONTH = moment()
  .add(1, 'M')
  .month();

export const WEEK_DATES = {
  0: 'Chủ Nhật',
  1: 'Thứ Hai',
  2: 'Thứ Ba',
  3: 'Thứ Tư',
  4: 'Thứ Năm',
  5: 'Thứ Sáu',
  6: 'Thứ Bảy',
};

export const WEEK_DAYS = {
  sunday: 'CN',
  monday: 'T2',
  tuesday: 'T3',
  wednesday: 'T4',
  thursday: 'T5',
  friday: 'T6',
  saturday: 'T7',
};

export const MONTH_INDEX = {
  jan: 0,
  feb: 1,
  mar: 2,
  apr: 3,
  may: 4,
  jun: 5,
  jul: 6,
  aug: 7,
  sep: 8,
  oct: 9,
  nov: 10,
  dec: 11,
};

export const MONTH_NAME = {
  jan: 'Tháng 01',
  feb: 'Tháng 02',
  mar: 'Tháng 03',
  apr: 'Tháng 04',
  may: 'Tháng 05',
  jun: 'Tháng 06',
  jul: 'Tháng 07',
  aug: 'Tháng 08',
  sep: 'Tháng 09',
  oct: 'Tháng 10',
  nov: 'Tháng 11',
  dec: 'Tháng 12',
};

export const CALENDAR_WEEKS = 6;
export const DAYS_OF_WEEK = 7;
export const NUMBER_OF_YEAR_IN_ROW = 4;
export const NUMBER_OF_MONTH_IN_ROW = 3;
