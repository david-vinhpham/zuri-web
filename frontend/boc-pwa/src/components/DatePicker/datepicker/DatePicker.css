/**
 * BOC VN (http://www.bocvietnam.com/)
 *
 * Copyright Â© 2019-present BOCVN, LLC. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

@import '../../variables.css';

.datepicker {
  position: relative;
}

.datepicker > button {
  position: absolute;
  right: 0;
  top: 0;
}

.backdrop {
  position: fixed;
  display: flex;
  justify-content: center;
  align-items: center;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background: rgba(0, 0, 0, 0.4);
  z-index: var(--z-index-highest);
}
