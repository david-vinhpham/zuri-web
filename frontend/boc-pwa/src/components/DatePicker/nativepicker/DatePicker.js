import React, { Component } from 'react';
import { func, string, objectOf, any, bool } from 'prop-types';
import moment from 'moment';
import withStyles from 'isomorphic-style-loader/lib/withStyles';

import s from './DatePicker.css';

class DatePicker extends Component {
  static propTypes = {
    maxDate: objectOf(any).isRequired,
    minDate: objectOf(any).isRequired,
    onChange: func.isRequired,
    name: string.isRequired,
    value: string.isRequired,
    touched: bool.isRequired,
    setFieldError: func.isRequired,
    errorMsg: string.isRequired,
    setFieldTouched: func.isRequired,
    label: string.isRequired,
    validationMsg: string,
    requiredMsg: string,
    dateFormat: string.isRequired,
    displayFormat: string.isRequired,
    require: bool,
  };

  static defaultProps = {
    require: false,
    validationMsg: '',
    requiredMsg: '',
  };

  constructor(props) {
    super(props);
    const { value } = props;
    this.state = {
      date: moment(value).isValid() ? value : '',
      errors: {},
    };
  }

  handleOnChange = event => {
    event && event.preventDefault();
    const {
      name,
      onChange,
      setFieldTouched,
      minDate,
      maxDate,
      setFieldError,
      validationMsg,
      dateFormat,
    } = this.props;
    const { value } = event.target;
    const currentDate = moment(value);
    const errors =
      currentDate > maxDate || currentDate < minDate
        ? { [name]: validationMsg }
        : null;
    this.setState({ date: currentDate, errors });
    onChange(name, !errors ? moment(value).format(dateFormat) : '');
    errors && setFieldError(name, errors[name]);
    setFieldTouched(name);
  };

  render() {
    const {
      minDate,
      maxDate,
      name,
      errorMsg,
      touched,
      label,
      requiredMsg,
      dateFormat,
      displayFormat,
      require,
    } = this.props;
    const { date, errors } = this.state;
    let resolveError =
      touched && (errors && errors[name]) ? errors[name] : errorMsg;
    resolveError = resolveError.includes(name) ? requiredMsg : resolveError;
    const calendarStyle =
      (errors && errors[name]) || (touched && resolveError)
        ? `${s.calendar} ${s.calendarError}`
        : s.calendar;
    const isStatusIcon = !errors && !errorMsg && moment(date).isValid();
    const labelStyle =
      date && moment(date).isValid() ? s.dateLabel : s.dateLabelInActive;

    return (
      <div className={s.datepicker}>
        <div className={calendarStyle}>
          <div className={labelStyle}>
            <span>
              {label}
              {require && <sup>(*)</sup>}
            </span>
          </div>
          <div className={s.dateCalendar}>
            <div>
              <i className={`material-icons ${s.icon}`}>date_range</i>
            </div>
            <div className={s.formControl}>
              <input
                type="date"
                name={name}
                value={moment(date).format(dateFormat)}
                onChange={this.handleOnChange}
                min={moment(minDate).format(dateFormat)}
                max={moment(maxDate).format(dateFormat)}
                onKeyDown={e => {
                  e.preventDefault();
                }}
              />
              <div className={s.formValue}>
                {moment(date).isValid()
                  ? moment(date).format(displayFormat)
                  : ''}
              </div>
            </div>
            <div className={s.statusIconWrapper}>
              {isStatusIcon && (
                <i className={`material-icons ${s.statusIcon}`}>check</i>
              )}
            </div>
          </div>
        </div>
        {touched && (
          <div className={s.errorMessage}>
            {resolveError.includes(name) ? requiredMsg : resolveError}
          </div>
        )}
      </div>
    );
  }
}

export default withStyles(s)(DatePicker);
