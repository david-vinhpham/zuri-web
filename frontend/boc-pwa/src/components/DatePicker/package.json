{
  "name": "date-picker",
  "version": "0.1.0",
  "private": true,
  "main": "index.js"
}
