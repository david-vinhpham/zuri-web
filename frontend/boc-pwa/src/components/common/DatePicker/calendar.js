import { chunk } from 'lodash';
import moment from 'moment';
import {
  THIS_MONTH,
  THIS_YEAR,
  CALENDAR_WEEKS,
  DAYS_OF_WEEK,
} from './constants';

export const zeroPad = (value, length) => `${value}`.padStart(length, '0');

export const getPreviousMonth = (month, year) => {
  const [previousMonth, previousYear] =
    +month > 1 ? [+month - 1, year] : [12, +year - 1];
  return { month: previousMonth, year: previousYear };
};

export const getNextMonth = (month, year) => {
  const [nextMonth, nextYear] =
    +month < 12 ? [+month + 1, year] : [1, +year + 1];
  return { month: nextMonth, year: nextYear };
};

export default (month = THIS_MONTH, year = THIS_YEAR) => {
  // First day and number of days in month
  const daysOfMonth = moment(`${year}-${month}`, 'YYYY-MM').daysInMonth();

  // Fill out day to 6 weeks
  const daysOfPreviousMonth = moment(`${year}-${zeroPad(month, 2)}`)
    .startOf('month')
    .day();
  const daysOfNextMonth =
    CALENDAR_WEEKS * 7 - (daysOfPreviousMonth + daysOfMonth);

  // Get the previous and next month and year
  const { month: previousMonth, year: previousYear } = getPreviousMonth(
    month,
    year,
  );
  const { month: nextMonth, year: nextYear } = getNextMonth(month, year);

  // Get number of days in previous month
  const previousMonthDays = moment(
    `${previousYear}-${previousMonth}`,
    'YYYY-MM',
  ).daysInMonth();

  // Build dates to be displayed from previous month
  const previousMonthDates = [
    ...Array.from({ length: daysOfPreviousMonth }, (date, index) => [
      previousYear,
      zeroPad(previousMonth, 2),
      zeroPad(index + 1 + (previousMonthDays - daysOfPreviousMonth), 2),
    ]),
  ];

  // Build dates to be displayed in current month
  const currentMonthDates = [
    ...Array.from({ length: daysOfMonth }, (date, index) => [
      year,
      zeroPad(month, 2),
      zeroPad(index + 1, 2),
    ]),
  ];

  // Build dates to be displayed from next month
  const nextMonthDates = [
    ...Array.from({ length: daysOfNextMonth }, (date, index) => [
      nextYear,
      zeroPad(nextMonth, 2),
      zeroPad(index + 1, 2),
    ]),
  ];

  // Combine all dates
  return chunk(
    [...previousMonthDates, ...currentMonthDates, ...nextMonthDates],
    DAYS_OF_WEEK,
  );
};
