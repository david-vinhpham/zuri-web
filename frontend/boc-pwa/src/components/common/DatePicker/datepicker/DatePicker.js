import React, { Component } from 'react';
import { string, func, bool } from 'prop-types';
import { noop } from 'lodash';
import moment from 'moment';
import withStyles from 'isomorphic-style-loader/lib/withStyles';
import { Input } from 'homecredit-ui';
import Calendar from '../calendar/Calendar';
import s from './DatePicker.css';

class DatePicker extends Component {
  static propTypes = {
    name: string.isRequired,
    touched: bool.isRequired,
    dateFormat: string.isRequired,
    displayFormat: string.isRequired,
    label: string,
    errorMsg: string,
    onChange: func,
    require: bool,
  };

  static defaultProps = {
    errorMsg: '',
    label: '',
    onChange: noop,
    require: false,
  };
  constructor(props) {
    super(props);
    this.state = {
      value: null,
      maxDate: null,
      minDate: null,
      isCalendarOpen: false,
    };
  }

  static getDerivedStateFromProps(props, state) {
    const { value, maxDate, minDate } = props;
    const {
      value: cachedValue,
      maxDate: cachedMaxDate,
      minDate: cachedMinDate,
    } = state;
    if (
      value !== cachedValue ||
      maxDate !== cachedMaxDate ||
      minDate !== cachedMinDate
    ) {
      return {
        value,
        maxDate,
        minDate,
      };
    }

    return null;
  }

  toggleCalendar = () =>
    this.setState({ isCalendarOpen: !this.state.isCalendarOpen });

  handleDateChange = date => {
    const { value } = this.state;
    const { onChange, name } = this.props;
    onChange(name, date);
    if (moment(date).diff(moment(value)) === 0) {
      this.setState({
        value: date,
      });
    }
  };

  render() {
    const {
      name,
      label,
      errorMsg,
      touched,
      require,
      dateFormat,
      displayFormat,
    } = this.props;
    const { value, minDate, maxDate, isCalendarOpen } = this.state;

    return (
      <>
        <div className={s.datepicker}>
          <div onClick={this.toggleCalendar}>
            <Input
              errorMsg={errorMsg}
              touched={touched}
              label={label}
              icon="date_range"
              name={name}
              readOnly
              type="text"
              value={
                (moment(value).isValid() &&
                  moment(value).format(displayFormat)) ||
                ''
              }
              require={require}
              msgPosition="below"
            />
          </div>
        </div>

        {isCalendarOpen && (
          <div className={s.backdrop}>
            <Calendar
              date={moment(value)}
              maxDate={maxDate}
              minDate={minDate}
              onDateChanged={this.handleDateChange}
              onClose={this.toggleCalendar}
              dateFormat={dateFormat}
            />
          </div>
        )}
      </>
    );
  }
}

export default withStyles(s)(DatePicker);
