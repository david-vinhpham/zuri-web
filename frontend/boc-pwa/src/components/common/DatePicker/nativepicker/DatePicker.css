/**
 * BOC VN (http://www.bocvietnam.com/)
 *
 * Copyright Â© 2019-present BOCVN, LLC. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

.datepicker {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  height: var(--b-space-4xl);
}

.calendar {
  border-bottom: var(--b-space-3xs) solid var(--c-ui-normal);
  padding-left: var(--b-space-xs);
}

.calendarError {
  transition: border 0.5s ease-out;
  border-bottom: var(--b-space-3xs) solid var(--c-brand);
}

.dateLabel {
  padding: var(--b-space-2xs) 0 0 var(--b-space-xl);
  font-size: calc(var(--font-size-tiny) * 1.2);
  color: var(--c-label);
  transition: all 0.2s ease;
}

.dateLabelInActive {
  transform: translate3d(var(--b-space-xl), calc(var(--b-space-md) * 1.5), 0);
  font-size: var(--font-size-small);
  color: var(--c-label);
  transition: transform 0.2s ease;
}

.dateCalendar {
  color: var(--c-label);
  display: flex;
  width: 100%;
  overflow: hidden;
}

.errorMessage {
  padding-top: var(--b-space-2xs);
  font-size: calc(var(--font-size-tiny) * 1.2);
  color: var(--c-brand);
}

.formControl {
  margin: var(--b-space-2xs) var(--b-space-xs);
  color: var(--c-label);
  width: 100%;
  display: flex;
  align-items: center;
}

.formValue {
  position: absolute;
  font-size: var(--font-size-small);
  z-index: 0;
}

input[type='date'] {
  position: absolute;
  background: transparent;
  color: transparent;
  display: -webkit-inline-flex;
  overflow: hidden;
  border: none;
  outline: none;
  user-select: none;
  height: calc(var(--b-space-xl) * 0.78);
  width: 75%;
  z-index: 1;
}

input[type='date']::-webkit-clear-button {
  display: none;
}

input[type='date']::-ms-clear {
  display: none;
}

input[type='date']::-webkit-inner-spin-button {
  display: none;
}

input[type='date']::-webkit-calendar-picker-indicator {
  position: absolute;
  width: 100%;
  height: 100%;
  opacity: 0;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
}

input[type='date']:focus {
  user-select: none;
  opacity: 0;
}

input[type='date']:hover {
  cursor: pointer;
}

.icon {
  color: var(--c-brand);
  margin-top: var(--b-space-2xs);
}

.statusIconWrapper {
  display: flex;
  justify-content: center;
  align-items: center;
}

.statusIcon {
  font-size: var(--font-size-big);
  color: var(--c-success);
}
