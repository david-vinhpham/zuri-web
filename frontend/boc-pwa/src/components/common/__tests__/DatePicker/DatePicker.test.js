/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';

import DatePicker from '../../DatePicker';

describe('DatePicker', () => {
  it('should render DatePicker with default props', () => {
    const wrapper = renderer.create(<DatePicker />);
    const snapshot = wrapper.toJSON();

    expect(snapshot).toMatchSnapshot();
  });
});
