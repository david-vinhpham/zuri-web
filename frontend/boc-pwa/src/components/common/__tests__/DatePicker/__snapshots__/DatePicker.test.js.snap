// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker should render DatePicker with default props 1`] = `
<div
  className="datepicker"
>
  <div
    onClick={[Function]}
  >
    <div
      className="Input_wrapper__WALTg Input_hasValue__-61ZX"
    >
      <i
        className="material-icons Input_icon__1jjsS"
      >
        date_range
      </i>
      <input
        accept=""
        autoComplete=""
        className="Input_input__1IXXH Input_inputIcon__1np_m"
        disabled={false}
        id=""
        maxLength={null}
        name=""
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        placeholder=""
        readOnly={true}
        type="text"
        value="2019-06-18T12:37:36+07:00"
      />
      <label
        className="Input_label__2ZxWx Input_inputIcon__1np_m"
        htmlFor=""
      >
        
      </label>
      <div
        className="Input_underline__Xwqb3"
      />
    </div>
  </div>
</div>
`;
